### PR TITLE
keep fixing warnings in 'core'

### PR DESCRIPTION
**Pull request by satorg**
For #486 (a follow-up to #569).

### CODE REVIEW

File: core/src/main/scala/cats/collections/Heap.scala [hunk 1]
```diff
@@ -308,11 +308,7 @@ object Heap {
     override def unbalanced: Boolean = size < (1L << height) - 1L
   }
 
-  private[collections] case object Leaf extends Heap[Nothing] {
-    def apply[A](): Heap[A] = this.asInstanceOf[Heap[A]]
-
-    def unapply[A](heap: Heap[A]): Boolean = heap.isEmpty
-
+  final private[collections] case class Leaf[A] private () extends Heap[A] {
```
Comment (satorg) [inline core/src/main/scala/cats/collections/Heap.scala:311]: Pretty much the same fix as it was in `Diet.scala` for the "match may not be exhaustive" warnings.

**Comment by satorg** [comment]
Interesting.. There are no tests reported as failed but rather
> FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory

https://github.com/typelevel/cats-collections/actions/runs/3751710955/jobs/6373035392#step:14:672

**Comment by armanbilge** [comment]
Right, that means the Node.js VM crashed. Maybe one of the generators is making too much data?

**Comment by satorg** [comment]
> Maybe one of the generators is making too much data?

Actually there was a mistake I made in the `Range.toIterator` implementation when was getting rid of `Stream`. The issue led to infinite loop in certain (quite rare) circumstances. A funny thing about it is that it was happening not in `RangeSuite` but rather in `DietSuite`.

However, when I started debugging it I noticed that there are similar issues with `Range.foreach` along with `foldLeft` and `foldRight`. The latter bug slipped through unnoticed because there were no tests that could reproduce necessary conditions for the bug to show up. So I decided to rewrite `RangeSuite` to make it using **scalacheck** and cover as many possible case as I could come up with.

File: core/src/main/scala/cats/collections/Range.scala [hunk 2]
```diff
@@ -103,14 +104,18 @@ final case class Range[A](val start: A, val end: A) {
...
     var i = start
-    while (order.lteqv(i, end)) {
+    // Visit values in range [start, end), i.e. excluding `end`.
+    while (order.lt(i, end)) {
       f(i)
       i = discrete.succ(i)
     }
+    // Visit the last (or the only) value, if any.
+    if (order.eqv(i, end))
+      f(i)
```
Comment (satorg) [inline core/src/main/scala/cats/collections/Range.scala:118]: The previous implementation was resulting to infinite loops in cases like
```
Range(123.toByte, Byte.MaxValue).foreach(_ => ())
```
I.e. due to integer overflows that could not be caught previously.

**Comment by satorg** [comment]
Hmm...
```
[error]  * the type hierarchy of object cats.collections.PairingHeap#Leaf is different in current version. Missing types {cats.collections.PairingHeap}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("cats.collections.PairingHeap$Leaf$")
```
not sure, why Mima gets triggered on `PairingHeap.Leaf` which is private and always has been that way...

Moreover, I did pretty much the same trick for `Diet` last time and there were no Mima errors about it.
@armanbilge do you have a clue by chance?

**Comment by armanbilge** [comment]
> not sure, why Mima gets triggered on `PairingHeap.Leaf` which is private and always has been that way...

MiMa has some issues sometimes with private classes inside objects. It should be safe to exclude.

**Comment by armanbilge** [comment]
Hmm, I see MiMa is complaining about many things. We've already discussed breaking bincompat in https://github.com/typelevel/cats-collections/issues/277#issuecomment-1315664123, maybe we should just bump the base version to 1.0?

**Comment by satorg** [comment]
@armanbilge 
> Hmm, I see MiMa is complaining about many things.

They are all about `PairingHeap.Leaf` and its internals. So seems I managed to calm Mima down with just two filters.

But yeah, it is quite weird that we have to have those filters for it after all. The only difference between `Diet.EmptyNode` and `PairingHeap.Leaf` I can spot is that the former's visibility is `private[collections]` wheres the latter's is just `private`. I can guess that the compiler may optimize private members out more aggressively comparing to package-local ones. But it's just my guess – not sure if I want to dig into it further :)

**Comment by armanbilge** [comment]
> former's visibility is `private[collections]` wheres the latter's is just `private`

Yes, I've noticed similar behavior too. My guess is that `private` is Java-style package private (i.e. non-public) but `private[package]` is Scala-private but Java-public. Which is why MiMa considers them differently.

**Review by gemelen — APPROVED** [review]

**Pull request merged**